src/pandaboxlib/pandaboxlib.py

```diff
@@ -6,80 +6,282 @@
+
+        """
+
+        def parse_fpga_from_idn(idn):
+            return idn.split(": ")[2]
+
+        # Read installed FPGA FW version
+        fpga_installed = parse_fpga_from_idn(self._read_idn())
+
+        # Read design FPGA FW version
+        offset = file.tell()
+        file.seek(0)
+        fpga_design = parse_fpga_from_idn(file.readline())
+        file.seek(offset)
+
+        # Validate
+        return fpga_design == fpga_installed
+
+    def save(self, path):
+        """Save design to file"""
+        with open(path, "wt") as file:
+
+            # Save firmware version
+            self._save_fw_version(file)
+
+            # Save the CONFIG state
+            self._save_state(file, '*CHANGES.ATTR?')
+            self._save_state(file, '*CHANGES.CONFIG?')
+
+            # Save individual tables
+            #
+            #   N.b. Must read complete table response before processing!
+            #
+            tables = list(self._read_response('*CHANGES.TABLE?'))
+            for table in tables:
+                self._save_table(file, table)
+
+            # Save metadata
+            #
+            #   N.b. Table and value results are mixed together
+            #
+            metadata = list(self._read_response('*CHANGES.METADATA?'))
+            for line in metadata:
+                self._save_metadata(file, line)
+
+    def load(self, path):
+        """Load design from file"""
+        with open(path,"rt") as file:
+
+            # Validate FPGA firmware version
+            if not self._validate_fw_version(file):
+                raise ValueError("Design and installed FPGA FW versions do not match")
+
+            # Apply design
+            for line in file:
+                self.sock.sendall(line.encode())
 
 
 class PandA:
 
-    def __init__(self, host):
+    sock = None
+
+    def __init__(self, host, port=8888):
+        """Initializer"""
         self.host = host
-        self.port = 8888
+        self.port = port
+
+    def __del__(self):
+        """Finalizer/destructor
+
+        Actions:
 
-        # Create a TCP socket
-        self.sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
-        self.sock.setsockopt(socket.SOL_TCP, socket.TCP_NODELAY, 1)
-        self.sock.settimeout(1)
+          * Close socket gracefully
+
+        """
+        if self.sock is not None:
+            self.disconnect_from_panda()
 
     def connect_to_panda(self):
-        self.sock.connect((self.host, self.port))
+        """Create socket connection to host"""
+        if self.sock is None:
+            self.sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+            self.sock.setsockopt(
+                socket.SOL_TCP,          # Disable Nagle algorithm
+                socket.TCP_NODELAY,      # (https://linux.die.net/man/7/tcp)
+                1                        # Is this really necessary?
+            )
+            self.sock.settimeout(1)
+            self.sock.connect((self.host, self.port))
```

Comment:
It might be nice to put a try-except around this, with a context message something like "unable to connect to panda at host {host} port {port}".

Reply:
Yep, agree. Socket exceptions are not super clear…